pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3132,8 +3132,20 @@ public String startReplaceSegments(String tableNameWithType, List<String> segmen
               // Update segment lineage entry to 'REVERTED'
               updateSegmentLineageEntryToReverted(tableNameWithType, segmentLineage, entryId, lineageEntry);
 
+              // Update segmentsTo of the 'REVERTED' entry to handle the following case:
```

Comment:
Can we merge this part with the `updateSegmentLineageEntryToReverted()`? In most of the retry cases, the `segmentsTo` will be the same for both upload, and we may directly remove the existing `IN_PROGRESS` entry instead of making it `REVERTED` but with empty `segmentsTo`

Reply:
I don't think we can directly remove the `IN_PROGRESS` entry, otherwise any segments in `segmentsTo` can be queried, while the next batch of segment upload is still in progress.
E.g. if the 1st lineage entry is like `{segmentsTo: [s3, s4]}` and the next one is `{segmentsTo: [s4, s5]}`, then s3 can be queried once the lineage is removed. So only when s3 is removed can we safely delete the old `IN_PROGRESS` entry.

Reply:
@jackjlli I think @Jackie-Jiang means we can remove the `IN_PROGRESS` entry if the segmentsTo are exactly the same.
Say the initial state is:
entry1: { segmentsFrom: [s1, s2], segmentsTo: [s3, s4], status: IN_PROGRESS}
Rerunning the protocol with same segmentsTo will remove entry1 and create entry2:
entry2: { segmentsFrom: [s1, s2], segmentsTo: [s3, s4], status: IN_PROGRESS}

Even if s3 and s4 are not deleted, they will be filtered out based on entry2.

Reply:
We can remove the old `IN_PROGRESS` entry only when all the `segmentsTo` are covered by the new `IN_PROGRESS` entry, thus none of these segments will be queryable. If that's not the case, we make it `REVERTED` and remove the common segments from the `segmentsTo` (the current logic in the PR)

Reply:
I see. In that case, the updated `segmentsTo` field is already empty, we can simply clean up that lineage entry. 👍

Reply:
If we only have partial overlapping for the segmentsTo, we still need to mark it as `REVERTED`. I've merged the logic into `updateSegmentLineageEntryToReverted()` and renamed as `deleteORUpdateSegmentLineageEntryToReverted()`. Updated the comments to list both examples to make it clear.